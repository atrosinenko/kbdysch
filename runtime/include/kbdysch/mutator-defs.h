#ifndef KBDYSCH_MUTATOR_DEFS_H
#define KBDYSCH_MUTATOR_DEFS_H

#include <stdint.h>

#ifdef __cplusplus
extern "C" {
#endif

// Common definitions for harness executables and mutator library

#define MUTATOR_SHM_VARS_ENV_NAME "__KBDYSCH_MUTATOR_SHM_VARS_ID"
#define MUTATOR_SHM_LOG_ENV_NAME "__KBDYSCH_MUTATOR_SHM_LOG_ID"

// Mutator and harness communicate by means of dedicated SHM areas.
// Assumptions:
// * Mutator should never crash, even on malformed SHM contents.
//   After all, if fuzzer finds a crash, it is quite expectably that
//   harness-writable memory can be corrupted.
// * Mutator library and harness executable should match exactly,
//   no forward/backward compatibility is assumed for now.
// * Each successful invocation of the same harness should produce the
//   same layout of variables SHM area. It is perfectly OK for different
//   harnesses or differently-configured harness to produce different layouts
//   * this area can be parsed once during fuzzing session warm-up and then
//     pointers can be remembered
//
// Variables area layout:
//
// N records are concatenated without any padding in between.
// N does not change across fuzzing session.
// Each record has the following layout:
// - struct mutator_var_header
// - (name_bytes) bytes with variable name (not necessary 0-terminated)
// - (bytes_per_element * max_num_elements) bytes of payload
// The last record should have type == MUTATOR_VAR_STOP.
//
// Then, at MUTATOR_SHM_VARS_BYTES offset, another area is located that
// is zeroed by the forkserver before each run. This temporary area has
// the same layout as the first one, but only a few supported fields are
// updated.
//
// Log area layout:
//
// HASH_CHARS bytes of input hash is followed by log records.
// Each harness run can produce an arbitrary number of log records
// concatenated without any padding in between.
// Each record has the following layout:
// - struct mutator_log_record_header
// - payload depending on the type field
// The last record should have type == MUTATOR_LOG_STOP.

enum mutator_fixed_record_type {
  MUTATOR_VAR_STOP = 0,
  MUTATOR_VAR_COUNTERS,
  MUTATOR_VAR_STRINGS,
};

enum mutator_opcode {
  MUTATOR_LOG_STOP = 0,
  MUTATOR_LOG_SET_OFFSET,
  MUTATOR_LOG_NEW_RES,
  MUTATOR_LOG_REF_RES,
  MUTATOR_LOG_PROPOSE_CHANGE,
};

enum mutator_resource_kind {
  RESOURCE_KIND_FD,
  RESOURCE_KIND_FILE_NAME,
  RESOURCE_KIND_HARNESS_SPECIFIC_1,

  MUTATOR_MAX_RESOURCE_KINDS
};

#define UNALIGNED __attribute__((aligned(1)))
typedef uint16_t mutator_num_elements_t UNALIGNED;
typedef uint64_t mutator_u64_var_t UNALIGNED;

#pragma pack(push,1)
struct mutator_var_header {
  uint16_t type;
  uint16_t name_bytes;
  uint16_t bytes_per_element;
  mutator_num_elements_t max_num_elements;

  mutator_num_elements_t num_elements_real;
};

struct mutator_log_record_header {
  uint16_t type;
  uint16_t size; // including this header
};
struct mutator_log_set_offset {
  uint32_t offset;
};
struct mutator_log_new_resource {
  uint8_t kind;
  uint8_t id;
#define MUTATOR_MAX_RESOURCE_IDS 256
};
struct mutator_log_ref_resource {
  uint8_t kind;
  uint8_t id_bytes;
  uint8_t id;
  uint32_t offset;
};
struct mutator_log_propose_change {
  uint32_t offset;
  uint32_t size; // from 1 to 8 bytes
  uint64_t replacement;
};
#pragma pack(pop)

#define MUTATOR_MAX_TEST_CASE_LENGTH (1 << 20)

#define MUTATOR_MAX_VARIABLES 100
#define MUTATOR_SHM_VARS_BYTES 4000
#define MUTATOR_SHM_VAR_IN_CURRENT_AREA(ptr) \
    ((void *)(((uint8_t *)(ptr)) + MUTATOR_SHM_VARS_BYTES))

#define MUTATOR_MAX_LOG_BYTES 4000
#define MUTATOR_SHM_LOG_BYTES (HASH_CHARS + MUTATOR_MAX_LOG_BYTES)
#define MUTATOR_MAX_LOG_ENTRIES(type) (MUTATOR_MAX_LOG_BYTES / (sizeof(struct mutator_log_record_header) + sizeof(struct type)))
#define MUTATOR_MAX_OFFSETS MUTATOR_MAX_LOG_ENTRIES(mutator_log_set_offset)

#ifdef __cplusplus
}
#endif

#endif  // KBDYSCH_MUTATOR_DEFS_H
