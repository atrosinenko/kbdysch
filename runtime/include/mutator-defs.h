#ifndef KBDYSCH_MUTATOR_DEFS_H
#define KBDYSCH_MUTATOR_DEFS_H

// Common definitions for harness executables and mutator library

#define MUTATOR_ENV_NAME "__KBDYSCH_MUTATOR_SHM_ID"

#define MUTATOR_MAX_TEST_CASE_LENGTH (1 << 20)
#define MUTATOR_MAX_TRIM_OFFSETS 128
#define MUTATOR_MAX_SHM_WORDS 1024

#define MUTATOR_OP_STOP 0
#define MUTATOR_OP_TRIM_OFFSET 1

#endif  // KBDYSCH_MUTATOR_DEFS_H
