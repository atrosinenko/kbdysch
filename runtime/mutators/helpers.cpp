#include "helpers.h"

#include "kbdysch/options.h"

#include <errno.h>
#include <fcntl.h>
#include <stdlib.h>
#include <string.h>
#include <sys/shm.h>
#include <unistd.h>

DECLARE_BOOL_KNOB(debug_logging, "KBDYSCH_MUTATOR_DEBUG")

FILE *kbdysch::mutator::error_log;

FILE *kbdysch::mutator::create_temp_file(
    const char *prefix, const char *suffix) {
  char log_name[128];
  sprintf(log_name, "%s%d%s", prefix, getpid(), suffix);
  return fopen(log_name, "w");
}

void kbdysch::mutator::init_error_logging() {
  if (!debug_logging)
    return;

  error_log = create_temp_file("/tmp/kbdysch-mutator-", ".log");
  setvbuf(error_log, NULL, _IONBF, 0);
}

void kbdysch::mutator::deinit_error_logging() {
  if (error_log)
    fclose(error_log);
}

bool kbdysch::mutator::in_mem_area_bounds(
    const void *mem_area, size_t mem_size, const void *ptr, size_t requested_size) {
  uintptr_t mem_start = (uintptr_t)mem_area;
  uintptr_t ptr_start = (uintptr_t)ptr;

  if ((uintptr_t)ptr < (uintptr_t)mem_area) {
    ERR("Pointer %p outside of memory area starting at %p.\n",
        ptr, mem_area);
    return false;
  }
  size_t offset_in_mem_area = (uintptr_t)ptr - (uintptr_t)mem_area;
  if (offset_in_mem_area > mem_size ||
      requested_size > mem_size ||
      offset_in_mem_area + requested_size > mem_size) {
    ERR("Pointer %p at invalid offset 0x%zx in memory area of size 0x%zx.\n",
        ptr, offset_in_mem_area, mem_size);
    return false;
  }
  return true;
}

namespace kbdysch {
namespace mutator {

shm_segment::shm_segment(const char *env_var_name,
                         size_t allocatable_bytes, size_t total_bytes) {
  char id_str[32];
  // Allocate SHM segment
  ShmId = shmget(IPC_PRIVATE, total_bytes, 0600);
  void *shm_segment = shmat(ShmId, NULL, 0);
  AllocatableBuffer = buffer_ref(shm_segment, allocatable_bytes);
  // Make SHM discoverable by harness
  sprintf(id_str, "%d", ShmId);
  setenv(env_var_name, id_str, 1);
}

shm_segment::~shm_segment() {
  shmdt(AllocatableBuffer.bytes());
  shmctl(ShmId, IPC_RMID, NULL);
}

temp_dir::temp_dir(const char *name) {
  strncpy(dir_name, name, MAX_DIR_NAME);
  dir_name[MAX_DIR_NAME - 1] = 0;
  if (!mkdtemp(dir_name))
    FATAL("Cannot create temporary directory %s: %s", dir_name, strerror(errno));

  dir_fd = open(dir_name, O_RDONLY);
  if (dir_fd < 0)
    FATAL("Cannot open directory %s: %s", dir_name, strerror(errno));
}

temp_dir::~temp_dir() {
  close(dir_fd);
  // TODO remove directory
}

ssize_t temp_dir::read_file(const char *name, buffer_ref buf) {
  int file_fd = openat(dir_fd, name, O_RDONLY);
  if (file_fd < 0) {
    ERR("Cannot open file %s for reading: %s\n", name, strerror(errno));
    return -1;
  }

  ssize_t length = read(file_fd, buf.bytes(), buf.size());
  if (length < 0)
    FATAL("Cannot read file %s: %s", name, strerror(errno));

  close(file_fd);
  return length;
}

void temp_dir::write_file(const char *name, buffer_ref buf) {
  int file_fd = openat(dir_fd, name, O_WRONLY | O_CREAT | O_TRUNC, 0644);
  if (file_fd < 0)
    FATAL("Cannot open file %s for writing: %s.\n", name, strerror(errno));
  ssize_t length = write(file_fd, buf.bytes(), buf.size());
  if (length != buf.size())
    FATAL("Cannot write %zu bytes to file %s: %s.\n",
          buf.size(), name, strerror(errno));
  close(file_fd);
}

} // namespace mutator
} // namespace kbdysch
