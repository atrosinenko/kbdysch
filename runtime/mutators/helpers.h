#ifndef KBDYSCH_MUTATORS_HELPERS_H
#define KBDYSCH_MUTATORS_HELPERS_H

#include <stddef.h>
#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>

#include <array>
#include <vector>

namespace kbdysch {
namespace mutator {

extern FILE *error_log;
void init_error_logging();
void deinit_error_logging();

#define DECL_WITH_TYPE(type, new_name, ptr) \
  type *new_name = (type *)(ptr)

#define FATAL(fmt, ...)                                                        \
  do {                                                                         \
    fprintf(stderr, "MUTATOR: %s: %d: " fmt, __FILE__, __LINE__, __VA_ARGS__); \
    abort();                                                                   \
  } while (0)

#define ERR(...)                                         \
  do {                                                   \
    if (kbdysch::mutator::error_log)                     \
      fprintf(kbdysch::mutator::error_log, __VA_ARGS__); \
  } while (0)

#define DEBUG_TRACE_FUNC // fprintf(stderr, "MUTATOR: %s: called\n", __func__)
#define DEBUG(fmt, ...)  // fprintf(stderr, "MUTATOR: %s: " fmt, __func__, __VA_ARGS__)

bool in_mem_area_bounds(const void *mem_area, size_t mem_size,
                        const void *ptr, size_t requested_size);

class buffer_ref {
public:
  buffer_ref()
      : Ptr(0), Size(0) {}

  buffer_ref(void *ptr, size_t size)
      : Ptr((uint8_t *)ptr), Size(size) {}

  template <typename T, std::size_t N>
  buffer_ref(std::array<T, N> &Array)
      : buffer_ref(Array.data(), Array.size()) {}

  template <typename T>
  buffer_ref(std::vector<T> &Vector)
      : buffer_ref(Vector.data(), Vector.size()) {}

  uint8_t *bytes() { return Ptr; }
  const uint8_t *bytes() const { return Ptr; }

  size_t size() const { return Size; }

  bool contains(const void *other_ptr, size_t requested_size) const {
    return in_mem_area_bounds(Ptr, Size, other_ptr, requested_size);
  }

  buffer_ref subbuf(size_t offset, size_t sub_size) {
    return buffer_ref(&Ptr[offset], sub_size);
  }
  const buffer_ref subbuf(size_t offset, size_t sub_size) const {
    return buffer_ref(&Ptr[offset], sub_size);
  }

private:
  uint8_t *Ptr;
  size_t Size;
};

// Array-backed buffer that has a statically-set maximum capacity but
// its logical size can vary from 0 to MaxSize
template <size_t MaxSize>
class array_buffer {
public:
  array_buffer()
      : Size(0) {}

  buffer_ref as_storage() {
    return buffer_ref(Storage);
  }
  const buffer_ref as_storage() const {
    return buffer_ref(Storage);
  }

  buffer_ref as_data() {
    return buffer_ref(Storage).subbuf(0, Size);
  }
  const buffer_ref as_data() const {
    return buffer_ref(Storage).subbuf(0, Size);
  }

  uint8_t *bytes() { return Storage.data(); }
  const uint8_t *bytes() const { return Storage.data(); }

  size_t size() const { return Size; }
  size_t capacity() const { return Storage.size(); }

  void resize(size_t new_size) {
    Size = new_size;
  }

  void memcpy_back(buffer_ref buffer) {
    memcpy(&Storage[Size], buffer.bytes(), buffer.size());
    Size += buffer.size();
  }

  buffer_ref subbuf(size_t offset, size_t sub_size) {
    return buffer_ref(Storage).subbuf(offset, sub_size);
  }
  const buffer_ref subbuf(size_t offset, size_t sub_size) const {
    return buffer_ref(Storage).subbuf(offset, sub_size);
  }

private:
  std::array<uint8_t, MaxSize> Storage;
  size_t Size;
};

inline bool buffer_contains(const buffer_ref buffer,
                            const void *ptr, size_t requested_size) {
  return buffer.contains(ptr, requested_size);
}

class shm_segment {
public:
  shm_segment(const char *env_var_name,
              size_t allocatable_bytes, size_t total_bytes);

  uint8_t *begin() {
    return AllocatableBuffer.bytes();
  }
  const uint8_t *begin() const {
    return AllocatableBuffer.bytes();
  }

  bool in_bounds(const void *ptr, size_t length) const {
    return AllocatableBuffer.contains(ptr, length);
  }

  ~shm_segment();

private:
  int ShmId;
  buffer_ref AllocatableBuffer;
};

class temp_dir {
public:
  temp_dir(const char *name);

  ssize_t read_file(const char *name, buffer_ref buf);
  void write_file(const char *name, buffer_ref buf);

  ~temp_dir();

private:
  static const size_t MAX_DIR_NAME = 128;
  char dir_name[MAX_DIR_NAME];
  int dir_fd;
};

} // namespace mutator
} // namespace kbdysch

#endif // KBDYSCH_MUTATORS_HELPERS_H
