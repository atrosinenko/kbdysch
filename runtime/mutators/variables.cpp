#include "variables.h"
#include "kbdysch/mutator-defs.h"
#include "helpers.h"

#include <algorithm>
#include <cmath>
#include <functional>
#include <utility>
#include <vector>

namespace kbdysch {
namespace mutator {

class counter_variable : public variable {
public:
  counter_variable(std::string &&name,
                   unsigned max_num_elements,
                   mutator_num_elements_t *num_elements_real,
                   mutator_u64_var_t *counters,
                   mutator_u64_var_t *aux_counters)
      : variable(name, max_num_elements, num_elements_real),
        Counters(counters), AuxCounters(aux_counters) {
    Accumulator.assign(max_num_elements, 0);
  }

  void accumulate() override;
  void print_scalar(FILE *stream, unsigned subscript) override;

  uint64_t get(unsigned index) const { return Counters[index]; }
  uint64_t get_accumulated(unsigned index) const { return Accumulator[index]; }

private:
  mutator_u64_var_t *Counters;
  mutator_u64_var_t *AuxCounters;
  std::vector<uint64_t> Accumulator;
};

void counter_variable::accumulate() {
  for (unsigned i = 0; i < Accumulator.size(); ++i)
    Accumulator[i] += AuxCounters[i];
}

void counter_variable::print_scalar(FILE *stream, unsigned subscript) {
  fprintf(stream, "%lu\t%lu",
          (unsigned long)Counters[subscript],
          (unsigned long)Accumulator[subscript]);
}

class string_variable : public variable {
public:
  string_variable(std::string &&name,
                  unsigned max_num_elements,
                  mutator_num_elements_t *num_elements_real,
                  std::vector<char *> &&strings,
                  unsigned max_string_length)
      : variable(name, max_num_elements, num_elements_real),
        Strings(strings), MaxStringLength(max_string_length) {}

  void accumulate() override {}

  void print_scalar(FILE *stream, unsigned subscript) override;

private:
  std::vector<char *> Strings;
  unsigned MaxStringLength;
};

void string_variable::print_scalar(FILE *stream, unsigned subscript) {
  char *str = Strings[subscript];
  str[MaxStringLength - 1] = '\0';
  fprintf(stream, "%s", str);
}

class success_rate_variable : public variable {
public:
  success_rate_variable(std::string &&name,
                        string_variable *labels,
                        counter_variable *success,
                        counter_variable *failure)
      : variable(name, 0, 0), Labels(labels), Success(success), Failure(failure) {}

  void accumulate() override;

  void print(FILE *stream, unsigned var_index) override;

  ~success_rate_variable();

protected:
  void print_scalar(FILE *stream, unsigned subscript) override {
    abort(); // Not implemented
  }

private:
  void print_label_with_rank(FILE *stream,
                             unsigned index_in_permutation,
                             std::function<double(unsigned)> rank);

  void print_summary(FILE *stream,
                     const std::string &description,
                     std::function<double(unsigned)> rank);

  string_variable *Labels;
  counter_variable *Success;
  counter_variable *Failure;

  std::vector<unsigned> Permutation; // scratch variable

  const unsigned NumTopElements = 10;
};

void success_rate_variable::accumulate() {
  Labels->accumulate();
  Success->accumulate();
  Failure->accumulate();
}

void success_rate_variable::print_label_with_rank(
    FILE *stream, unsigned index_in_permutation, std::function<double(unsigned)> rank) {
  unsigned index = Permutation[index_in_permutation];
  Labels->print_scalar(stream, index);
  fprintf(stream, " (%.2f)", rank(index));
}

void success_rate_variable::print_summary(
    FILE *stream,
    const std::string &description,
    std::function<double(unsigned)> rank) {
  unsigned size = Labels->num_elements_real();

  Permutation.resize(size);
  for (unsigned i = 0; i < size; ++i)
    Permutation[i] = i;

  std::sort(
      Permutation.begin(), Permutation.end(),
      [rank](unsigned a, unsigned b) {
        double ra = rank(a);
        double rb = rank(b);
        if (std::isnan(ra))
          ra = -1;
        if (std::isnan(rb))
          rb = -1;
        return ra < rb;
      });

  fprintf(stream, "  - %s:\n", description.c_str());
  if (size <= 2 * NumTopElements) {
    // A (1.00), B (1.23), C (3.45), X (99.85)
    fprintf(stream, "    ");
    for (unsigned i = 0; i < size; ++i) {
      print_label_with_rank(stream, i, rank);
      if (i + 1 < size)
        fprintf(stream, ", ");
    }
    fprintf(stream, "\n");
  } else {
    // A (1.01), B (1.23), ...
    // ..., U (95.12), X (99.85)
    fprintf(stream, "    ");
    for (unsigned i = 0; i < NumTopElements; ++i) {
      print_label_with_rank(stream, i, rank);
      fprintf(stream, ", ");
    }
    fprintf(stream, "...\n");
    fprintf(stream, "    ...");
    for (unsigned i = size - NumTopElements; i < size; ++i) {
      fprintf(stream, ", ");
      print_label_with_rank(stream, i, rank);
    }
    fprintf(stream, "\n");
  }
}

void success_rate_variable::print(FILE *stream, unsigned var_index) {
  auto total = [this](unsigned i) {
    return 0.0 + Success->get(i) + Failure->get(i);
  };
  auto total_acc = [this](unsigned i) {
    return 0.0 + Success->get_accumulated(i) + Failure->get_accumulated(i);
  };
  fprintf(stderr, "Variable #%u: %s\n", var_index, Name.c_str());
  print_summary(stream, "Usage count (queue)", total_acc);
  print_summary(stream, "Usage count (all)", total);
  print_summary(stream, "Success (queue)", [this, total_acc](unsigned i) {
    return Success->get_accumulated(i) / total_acc(i);
  });
  print_summary(stream, "Success (all)", [this, total](unsigned i) {
    return Success->get(i) / total(i);
  });
}

success_rate_variable::~success_rate_variable() {
  delete Labels;
  delete Success;
  delete Failure;
}

bool variable::create_from_shm(std::vector<variable *> &variables,
                               buffer_ref main_area,
                               buffer_ref aux_area,
                               uint8_t **ptr) {
  if (!main_area.contains(*ptr, sizeof(struct mutator_var_header)))
    return false;

  DECL_WITH_TYPE(struct mutator_var_header, header, *ptr);
  unsigned type = header->type;
  unsigned name_bytes = header->name_bytes;
  unsigned bytes_per_element = header->bytes_per_element;
  unsigned max_num_elements = header->max_num_elements;

  // Prevent occasional integer overflows
  if (name_bytes > 1024 ||
      bytes_per_element > 1024 ||
      max_num_elements > 1024)
    return false;

  size_t size = sizeof(*header) + name_bytes + bytes_per_element * max_num_elements;
  if (!main_area.contains(*ptr, size))
    return false;

  const char *name_ptr = (const char *)(*ptr + sizeof(*header));
  std::string name(name_ptr, name_bytes);
  while (!name.empty() && name.back() == '\0')
    name.pop_back();

  uint8_t *payload = *ptr + sizeof(*header) + name_bytes;
  uint8_t *aux_payload = aux_area.bytes() + (payload - main_area.bytes());

  *ptr += size;

  switch (type) {
  case MUTATOR_VAR_COUNTERS:
    if (bytes_per_element != 8) {
      DEBUG("Counter %s: unexpected bytes_per_element = %u.\n",
            name.c_str(), bytes_per_element);
      return false;
    }

    variables.push_back(new counter_variable(
        std::move(name), max_num_elements, &header->num_elements_real,
        (mutator_u64_var_t *)payload,
        (mutator_u64_var_t *)aux_payload));

    return true;
  case MUTATOR_VAR_STRINGS: {
    std::vector<char *> strings;
    strings.reserve(max_num_elements);
    char *str = (char *)payload;
    for (unsigned i = 0; i < max_num_elements; ++i, str += bytes_per_element)
      strings.push_back(str);

    variables.push_back(new string_variable(
        std::move(name), max_num_elements, &header->num_elements_real,
        std::move(strings), bytes_per_element));

    return true;
  }
  case MUTATOR_VAR_STOP:
    return false;
  case MUTATOR_VAR_SUCCESS_RATE: {
    if (variables.size() < 3)
      return false;
    auto failure = variables.back();
    variables.pop_back();
    auto success = variables.back();
    variables.pop_back();
    auto labels = variables.back();
    variables.pop_back();

    variables.push_back(new success_rate_variable(
        std::move(name),
        static_cast<string_variable *>(labels),
        static_cast<counter_variable *>(success),
        static_cast<counter_variable *>(failure)));
    return true;
  }
  default:
    DEBUG("Unknown var record type: 0x%x\n", type);
    return false;
  }
}

unsigned variable::num_elements_real() {
  unsigned num_elements = *NumElementsReal;
  if (num_elements > MaxNumElements) {
    fprintf(stderr, "!!! Too many elements: %u, using only the first %u ones.\n",
            num_elements, MaxNumElements);
    num_elements = MaxNumElements;
  }
  return num_elements;
}

void variable::print(FILE *stream, unsigned var_index) {
  unsigned num_elements = num_elements_real();
  if (num_elements == 1) {
    fprintf(stream, "Variable #%u:\t", var_index);
    print_scalar(stream, 0);
    fprintf(stream, "\t - %s\n", Name.c_str());
  } else {
    fprintf(stream, "Variable #%u: %s\n", var_index, Name.c_str());
    for (unsigned i = 0; i < num_elements; ++i) {
      fprintf(stream, " - [%d]\t", i);
      print_scalar(stream, i);
      fprintf(stream, "\n");
    }
  }
}

} // namespace mutator
} // namespace kbdysch
