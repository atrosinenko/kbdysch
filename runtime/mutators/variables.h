#ifndef KBDYSCH_MUTATORS_VARIABLES_H
#define KBDYSCH_MUTATORS_VARIABLES_H

#include "kbdysch/mutator-defs.h"
#include "helpers.h"

#include <string>

namespace kbdysch {
namespace mutator {

class variable {
public:
  // On success, returns true, updates `variables` and sets `*Ptr` to the next variable to be read.
  // On error, returns false and `*Ptr` is unspecified.
  static bool create_from_shm(std::vector<variable *> &variables,
                              buffer_ref main_area,
                              buffer_ref aux_area,
                              uint8_t **ptr);

  virtual void accumulate() = 0;
  virtual void print(FILE *stream, unsigned var_index);
  unsigned num_elements_real();

  virtual ~variable() {}

protected:
  variable(const std::string &name, unsigned max_num_elements, mutator_num_elements_t *num_elements_real)
      : Name(name), MaxNumElements(max_num_elements), NumElementsReal(num_elements_real) {}

  virtual void print_scalar(FILE *stream, unsigned subscript) = 0;

  std::string Name;

private:
  unsigned MaxNumElements;
  mutator_num_elements_t *NumElementsReal;
};

} // namespace mutator
} // namespace kbdysch

#endif // KBDYSCH_MUTATORS_VARIABLES_H
