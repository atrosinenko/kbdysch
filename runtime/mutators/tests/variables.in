[BEGIN:CALL DEINIT]
[END:FINI DONE]
Input
