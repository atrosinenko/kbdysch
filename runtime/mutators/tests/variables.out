Variable #0:    0   0        - Number of %-prefixed tags processed
Variable #1:    0   0        - File names reused
Variable #2:    0   0        - File names created
Variable #3:    0   0        - Valid FDs
Variable #4:    0   0        - Invalid FDs
Variable #5:    0   0        - Unique FDs
